### DISCUSSION TITLE
GUI integration

### DISCUSSION THREAD

**Discussion by CatCode79** [Davidster/ikari#17] [General]
It's not easy to choose a GUI in rust, at least for me.

[egui](https://github.com/emilk/egui): I found it difficult to manage the placement and size of the various elements
[kayak_ui](https://github.com/StarArawn/kayak_ui): I love the aesthetics and the use of nine-patches, the problem is that it only works with bevy..
[fyrox_ui](https://github.com/FyroxEngine/Fyrox/tree/master/fyrox-ui): it would be ideal (however to integrate with wgpu), but I read that it has performance problems
[thyme](https://github.com/Grokmoo/thyme): just an honorable mention, I don't think that's what we're after

In the end I would choose [iced](https://github.com/iced-rs/iced):
It has an integration with wgpu and [wgpu_glyph](https://github.com/hecrj/wgpu_glyph) and from the examples it seems to me that the widgets are customizable in style, I've never tried it so I have no idea if it could be the ideal solution or not.

What do you think? Did you already have a particular gui in mind?

**Comment by Davidster** [comment]
very interested in talking about this! it might take me a few days to find the time to research it. from my first impressions, egui and iced both look very cool! I would also probably aim towards iced due to the declarative/reactive api

**Comment by CatCode79** [comment]
Perhaps iced, not being immediate mode like egui, may have disadvantages in gamedev (where everything is usually redrawn at every frames). But I see that the gui was born within the Coffee game engine so perhaps my doubts are unfounded. Unfortunately I don't know exactly the pros and cons of the various gui technologies in the gamedev area.

In the egui repo there is the Disadvantages of immediate mode section which lists some of the defects I encountered, it is still a very good gui and it is gaining momentum, it is not said that in the future some disadvantages will disappear, but for now...

In the last newsletter of "This week in Rust" there was this interesting link:
https://kas-gui.github.io/blog/state-of-GUI-2022.html

I don't know if you know him, but if you have time follow Raph Levien, he is doing research work for a rust gui and knows a lot about it:
https://www.youtube.com/watch?v=zVUTZlNCb8U
https://raphlinus.github.io/ui/graphics/gpu/2021/10/22/swapchain-frame-pacing.html

Anyway picking a Gui in Rust isn't easy, so let's take our time, we can continue to implement rendering stuff even without the Gui for a while

**Reply by Davidster** [reply]
Yeah I agree with you that it's not clear how the tradeoffs will look for this project. Maybe we can just try to implement the FPS counter with both crates (iced and egui) and see how it goes. 

State of GUI article and looks cool! adding it to my list. I had the frame pacing article on my reading list too lol, haven't had a chance to check it out yet. 

I watched about half of the talk from Raph Levien last week. It was pretty interesting but I found it didn't go into the level of detail that I had hoped, more just an overview of the crates that exist right now.
